flocker/testtools/__init__.py

```diff
@@ -192,33 +194,61 @@ def assert_not_equal_comparison(case, a, b):
             "Expected a and b to be not-equal: " + "; ".join(messages))
 
 
-def loop_until(predicate):
+def function_serializer(function):
+    try:
+        return {
+            "function": str(function),
+            "file": getfile(function),
+            "line": getsourcelines(function)[1]
+        }
+    except IOError:
+        # One debugging method involves changing .py files and is incompatible
+        # with inspecting the source.
+        return {
+            "function": function,
+        }
+
+LOOP_UNTIL_ACTION = ActionType(
+    action_type="flocker:testtools:loop_until",
+    startFields=[Field("predicate", function_serializer)],
+    successFields=[Field("result", serializer=safe_repr)],
+    description="Looping until predicate is true.")
+
+LOOP_UNTIL_ITERATION_MESSAGE = MessageType(
+    message_type="flocker:testtools:loop_until:iteration",
+    fields=[Field("result", serializer=safe_repr)],
+    description="Predicate failed, trying again.")
+
+
+def loop_until(predicate, reactor=reactor):
     """Call predicate every 0.1 seconds, until it returns something ``Truthy``.
 
     :param predicate: Callable returning termination condition.
     :type predicate: 0-argument callable returning a Deferred.
 
+    :param reactor: The reactor implemetnation to use to delay.
+    :type reactor: ``IReactorTime``.
+
     :return: A ``Deferred`` firing with the first ``Truthy`` response from
         ``predicate``.
     """
-    try:
-        msg("Looping on %s (%s:%s)" % (predicate, getfile(predicate),
-                                       getsourcelines(predicate)[1]))
-    except IOError:
-        # One debugging method involves changing .py files and is incompatible
-        # with inspecting the source.
-        msg("Looping on %s" % (predicate,))
+    action = LOOP_UNTIL_ACTION(predicate=predicate)
 
-    d = maybeDeferred(predicate)
+    d = action.run(DeferredContext, maybeDeferred(action.run, predicate))
 
     def loop(result):
         if not result:
-            d = deferLater(reactor, 0.1, predicate)
+            LOOP_UNTIL_ITERATION_MESSAGE(
+                message_type="flocker:testtools:loop_until:iteration",
```

Comment:
`message_type` already appears in `LOOP_UNTIL_ITERATION_MESSAGE`. Does it need to be repeated here?